csharp/src/Apache.Arrow.Flight.Sql/Client/FlightSqlClient.cs

```diff
@@ -0,0 +1,1211 @@
+    /// <returns>A task that represents the asynchronous operation. The task result contains the FlightInfo describing where to access the dataset.</returns>
+    public Task<FlightInfo> GetFlightInfoAsync(FlightDescriptor descriptor)
+    {
+        var options = new FlightCallOptions();
+        return GetFlightInfoAsync(options, descriptor);
+    }
+
+    /// <summary>
+    /// Perform the indicated action, returning an iterator to the stream of results, if any.
+    /// </summary>
+    /// <param name="options">Per-RPC options</param>
+    /// <param name="action">The action to be performed</param>
+    /// <returns>An async enumerable of results</returns>
+    public async IAsyncEnumerable<FlightResult> DoActionAsync(FlightCallOptions options, FlightAction action)
+    {
+        if (options is null)
+            throw new ArgumentNullException(nameof(options));
+
+        if (action is null)
+            throw new ArgumentNullException(nameof(action));
+
+        var call = _client.DoAction(action, options.Headers);
+
+        await foreach (var result in call.ResponseStream.ReadAllAsync())
+        {
+            yield return result;
+        }
+    }
+
+    /// <summary>
+    /// Perform the indicated action with default options, returning an iterator to the stream of results, if any.
+    /// </summary>
+    /// <param name="action">The action to be performed</param>
+    /// <returns>An async enumerable of results</returns>
+    public async IAsyncEnumerable<FlightResult> DoActionAsync(FlightAction action)
+    {
+        await foreach (var result in DoActionAsync(new FlightCallOptions(), action))
+        {
+            yield return result;
+        }
+    }
+
+    /// <summary>
+    /// Get the result set schema from the server for the given query.
+    /// </summary>
+    /// <param name="options">Per-RPC options</param>
+    /// <param name="query">The UTF8-encoded SQL query</param>
+    /// <param name="transaction">A transaction to associate this query with</param>
+    /// <returns>The SchemaResult describing the schema of the result set</returns>
+    public async Task<Schema> GetExecuteSchemaAsync(FlightCallOptions options, string query,
+        Transaction? transaction = null)
+    {
+        transaction ??= NoTransaction();
+
+        if (options is null)
+            throw new ArgumentNullException(nameof(options));
+
+        if (string.IsNullOrEmpty(query))
+            throw new ArgumentException($"Query cannot be null or empty: {nameof(query)}");
+
+        FlightInfo schemaResult = null!;
+        try
+        {
+            var prepareStatementRequest = new ActionCreatePreparedStatementRequest { Query = query };
+            var action = new FlightAction(SqlAction.CreateRequest, prepareStatementRequest.PackAndSerialize());
+            var call = _client.DoAction(action, options.Headers);
+
+            await foreach (var result in call.ResponseStream.ReadAllAsync())
```

Comment:
Add `ConfigureAwait(false)` if supported (I've not tried on a foreach)

Reply:
done